python/ray/train/predictor.py

```diff
@@ -139,28 +165,53 @@ def predict(self, data: DataBatchType, **kwargs) -> DataBatchType:
         Args:
             data: A batch of input data of type ``DataBatchType``.
             kwargs: Arguments specific to predictor implementations. These are passed
-            directly to ``_predict_pandas``.
+            directly to ``_predict_numpy`` or ``_predict_pandas``.
 
         Returns:
             DataBatchType:
                 Prediction result. The return type will be the same as the input type.
         """
-        data_df = convert_batch_type_to_pandas(data, self._cast_tensor_columns)
-
         if not hasattr(self, "_preprocessor"):
             raise NotImplementedError(
                 "Subclasses of Predictor must call Predictor.__init__(preprocessor)."
             )
-
         if self._preprocessor:
-            data_df = self._preprocessor.transform_batch(data_df)
+            data = self._preprocessor.transform_batch(data)
+        try:
+            batch_format = TYPE_TO_ENUM[type(data)]
+        except KeyError:
+            raise RuntimeError(
+                f"Invalid input data type of {type(data)}, supported "
+                f"types: {list(TYPE_TO_ENUM.keys())}"
+            )
 
-        predictions_df = self._predict_pandas(data_df, **kwargs)
-        return convert_pandas_to_batch_type(
-            predictions_df,
-            type=TYPE_TO_ENUM[type(data)],
-            cast_tensor_columns=self._cast_tensor_columns,
-        )
+        has_predict_numpy = self.__class__._predict_numpy != Predictor._predict_numpy
+        has_predict_pandas = self.__class__._predict_pandas != Predictor._predict_pandas
+
+        if batch_format == BatchFormat.PANDAS:
+            if has_predict_pandas:
+                return self._predict_pandas(data, **kwargs)
+            elif has_predict_numpy:
+                predict_data = _convert_batch_type_to_numpy(data)
+                predictions = self._predict_numpy(predict_data, **kwargs)
+                return convert_batch_type_to_pandas(predictions)
+        elif batch_format == BatchFormat.NUMPY:
+            if has_predict_numpy:
+                return self._predict_numpy(data, **kwargs)
+            elif has_predict_pandas:
+                # Fallback to convert to pandas batch and call _predict_pandas
+                # ex: xgboost predict with np.ndarray batch data
+                predict_data = convert_batch_type_to_pandas(data)
+                predictions = self._predict_pandas(predict_data, **kwargs)
+                # Base predictor's return value are used by both BatchPredictor
+                # and Ray Serve, thus keep return value as raw DataFrame or Numpy
+                # without any TensorArray casting and leave it to caller to decide.
+                return convert_pandas_to_batch_type(
```

Comment:
this doesn't quite work because our dict with single key is only needed in prediction but not preprocessor, so the change lives locally in DLPredictor. As a result many of our existing notebooks will fail and we've gone back and forth a few times in previous commits.

How about we leave this output format as follow up PR instead of creeping in all possible changes of existing path?

Reply:
there should be no functional change right?

`_convert_batch_type_to_numpy` calls `convert_pandas_to_batch_type` underneath the hood-- the former is better just for readability purposes